hudi-common/src/main/java/org/apache/hudi/common/table/TableSchemaResolver.java

```diff
@@ -159,23 +167,67 @@ public Schema getTableAvroSchema() throws Exception {
+      schema = schemaFromCommitMetadata.get();
+    } else {
+      Option<Schema> schemaFromTableConfig = metaClient.getTableConfig().getTableCreateSchema();
+      if (schemaFromTableConfig.isPresent()) {
+        if (includeMetadataFields) {
+          schema = HoodieAvroUtils.addMetadataFields(schemaFromTableConfig.get(), hasOperationField);
+        } else {
+          schema = schemaFromTableConfig.get();
+        }
       } else {
-        return schemaFromTableConfig.get();
+        if (includeMetadataFields) {
+          schema = getTableAvroSchemaFromDataFile();
+        } else {
+          schema = HoodieAvroUtils.removeMetadataFields(getTableAvroSchemaFromDataFile());
+        }
       }
     }
-    if (includeMetadataFields) {
-      return getTableAvroSchemaFromDataFile();
-    } else {
-      return HoodieAvroUtils.removeMetadataFields(getTableAvroSchemaFromDataFile());
+
+    Option<String[]> partitionFieldsOpt = metaClient.getTableConfig().getPartitionFields();
+    if (metaClient.getTableConfig().isDropPartitionColumns()) {
+      schema = recreateSchemaWhenDropPartitionColumns(partitionFieldsOpt, schema);
+    }
+    return schema;
+  }
+
+  public static Schema recreateSchemaWhenDropPartitionColumns(Option<String[]> partitionFieldsOpt, Schema originSchema) {
+    // when hoodie.datasource.write.drop.partition.columns is true, partition columns can't be persisted in data files.
+    // And there are no partition schema if the schema is parsed from data files.
+    // Here we create partition Fields for this case, and use StringType as the data type.
+    Schema schema = originSchema;
+    if (partitionFieldsOpt.isPresent() && partitionFieldsOpt.get().length != 0) {
+      List<String> partitionFields = Arrays.asList(partitionFieldsOpt.get());
+
+      final Schema schema0 = originSchema;
+      boolean hasPartitionColNotInSchema = partitionFields.stream().anyMatch(
+          pt -> !HoodieAvroUtils.containsFieldInSchema(schema0, pt)
+      );
+      boolean hasPartitionColInSchema = partitionFields.stream().anyMatch(
+          pt -> HoodieAvroUtils.containsFieldInSchema(schema0, pt)
+      );
+      if (hasPartitionColNotInSchema && hasPartitionColInSchema) {
+        throw new HoodieIncompatibleSchemaException(
+            "Not support: Partial partition fields are still in the schema "
+                + "when enable hoodie.datasource.write.drop.partition.columns");
+      }
+
+      if (hasPartitionColNotInSchema) {
+        // when hasPartitionColNotInSchema is true and hasPartitionColInSchema is false, all partition columns
+        // are not in originSchema. So we create and add them.
+        List<Field> newFields = new ArrayList<>();
+        for (String partitionField: partitionFields) {
+          newFields.add(new Schema.Field(
+              partitionField, Schema.create(Schema.Type.STRING), "", JsonProperties.NULL_VALUE));
```

Comment:
Why do we assume that those partition fields would be strings?

Reply:
This is tricky, we should avoid that.

Reply:
if get table schema from data files and the setting `hoodie.datasource.write.drop.partition.columns` is enabled, we can't know what type it is. here is a little similar to `TimestampBasedKeyGenerator`, using `TimestampBasedKeyGenerator` will generate other partition values, and we also don't know the exact data type.

Reply:
how about store the partition type in hoodie.properties

Reply:
in some cases, the schema is out-of-date in hoodie.properties, even there is no schema stored in hoodie.properties.
I prefer to have a unified place to manage the latest schema, rather than the current logical that get schema from metadata, hoodie.properties, datafile in order.
Like deltalake and iceberg do.

Reply:
The point is that we're making forced conversion from the source type of the dataset to a string, which a) will likely break the ordering of such columns (when we convert, say, numbers to strings) b) might break queries expecting particular type of the column and now getting the string there

Reply:
@YannByron we should certainly unify schema management, but this conversation is beyond the point here -- we need to make sure we properly reconstruct types of the columns one way or the other

Reply:
i have try to use the origin data type of partition columns. See the codes above: if the partition columns is in the `originSchema` maybe extracted from `metaClient.getTableConfig().getTableCreateSchema()`, just return the `originSchema` without any operations. As i said, sometimes can't know what the real type if the `originSchema` is extract from data files.

Reply:
> i have try to use the origin data type of partition columns. See the codes above: if the partition columns is in the `originSchema` maybe extracted from `metaClient.getTableConfig().getTableCreateSchema()`, just return the `originSchema` without any operations. As i said, sometimes can't know what the real type if the `originSchema` is extract from data files.

That's exactly what we need to address. Discussed this w/ @xushiyan, for now we can explicitly call out that we only recommend this to be used by users who have their partition columns as strings, and in either 0.11.1 or 0.12 we can follow-up and properly reconstruct the original types. 

Let's create a ticket to follow-up on this.